README.md
# Bolsuite Python Connector
bolsuite.python obtiene las cotizaciones del mercado de capitales desde [Bolsuite](https://www.facebook.com/groups/274656386254166) dejando disposible la informacion realtime para ser utilizada desde cualquier programa Python.

Los metodos que provee bolsuite.python se dividen en 3 grupos:
- Paneles
- Profundidad de mercado (Nivel 2)
- Historico intradiario

Los tres grupos retornan pandas dataframes lo cual permite manejar la informacion de manera rapida, sencilla y en un "Pythonic way".

-----
## Guia Rapida

La aplicacion cuenta con un solo modulo que se llama `connector` el cual permite acceder a toda la informacion.

	import bolsuite as bs

	bsc = bs.Connector()
	# Si bolsuite no se esta ejecutando en la misma computadora o esta configurado 
	# en un puerto diferente se puede indicar en la URL como parametro
	# bsc = bs.Connector(base_url="http://10.20.43.20:6091")
	
	# Obtener todos los indices
	bsc.indices()
	
	# Obtener un indice filtrado por nombre
	bsc.indices(ticker='MERVAL')
	
	# Obtener todas las acciones lideres
	bsc.bluechips()

	# Obtener todas las acciones lideres filtrada por vencimiento
	bsc.bluechips(settlement='Spot') 

	# Obtener una accion del panel de lideres filtrada por nombre
	bsc.bluechips(ticker='PAMP')

	# Obtener una accion del panel de lideres filtrada por nombre y vencimiento
	bsc.bluechips(ticker='PAMP', settlement='24hs')
	
	# Obtener todas las acciones del panel general
	bsc.general_board()

	# Obtener todas las acciones del panel general filtrada por vencimiento
	bsc.general_board(settlement='48hs') 

	# Obtener una accion del panel general filtrada por nombre
	bsc.general_board(ticker='LOMA')

	# Obtener una accion del panel general filtrada por nombre y vencimiento
	bsc.general_board(ticker='LOMA', settlement='Spot')

	# Obtener todos los cedears
	bsc.cedear_stocks()

	# Obtener todos los cedears filtrados por vencimiento
	bsc.cedear_stocks(settlement='24hs') 

	# Obtener un cedear filtrado por nombre
	bsc.cedear_stocks(ticker='C')

	# Obtener un cedear filtrado por nombre y vencimiento
	bsc.cedear_stocks(ticker='C', settlement='48hs')
	
	# Obtener todos los bonos
	bsc.government_bonds()

	# Obtener todos los bonos filtrados por vencimiento
	bsc.government_bonds(settlement='Spot') 

	# Obtener un bono filtrado por nombre
	bsc.government_bonds(ticker='AO20')

	# Obtener un bono filtrado por nombre y vencimiento
	bsc.government_bonds(ticker='AO20', settlement='24hs')
	
	# Obtener todas las letras
	bsc.government_short_term_bonds()

	# Obtener todas las letras filtradss por vencimiento
	bsc.government_short_term_bonds(settlement='48hs') 

	# Obtener una letra filtrada por nombre
	bsc.government_short_term_bonds(ticker='S11M0')

	# Obtener una letra filtrada por nombre y vencimiento
	bsc.government_short_term_bonds(ticker='S11M0', settlement='Spot')
	
	# Obtener todas opciones de un subyacente
	bsc.options(underlying_asset='GGAL')
	
	# Obtener todas las cauciones
	bsc.repos()
	
	# Obtener todas las cauciones filtradas por dia
	bsc.repos(days=7)

	# Obtener todas las cauciones filtradas por moneda
	bsc.repos(currency='ARS')

	# Obtener una caucion filtrada por dia y por moneda
	bsc.repos(days=1, currency='USD')
	
	# Obtener todas las obligaciones negociables
	bsc.corporate_bonds()

	# Obtener todas las obligaciones negociables por vencimiento
	bsc.corporate_bonds(settlement='24hs') 

	# Obtener una obligacion negociable filtrada por nombre
	bsc.corporate_bonds(ticker='BFCPO')

	# Obtener una obligacion negociable filtrado por nombre y vencimiento
	bsc.corporate_bonds(ticker='BFCPO', settlement='48hs')
	
	# Obtener todos los ADRs
	bsc.adrs()
	
	# Obtener un ADR filtrado por nombre
	bsc.adrs(ticker='GGAL.O')
	
	# Obtener todos los pares de divisas
	bsc.currency_pairs()
	
	# Obtener un par de divisas filtrado por nombre
	bsc.currency_pairs(ticker='USD-ARS')
	
	# Obtener todos los pares de divisas
	bsc.commodities()
	
	# Obtener un ADR filtrado por nombre
	bsc.commodities(ticker='Brent Oil')
	
	# Obtener todos los activos del panel personal
	bsc.personal_portfolio()
	
	# Obtener un activo del panel personal filtrado por nombre
	bsc.personal_portfolio(ticker='COME')
		
	# Obtener la profundidad de mercado de los activos seleccionados
	bsc.level_2_quotes(tickers=['GFGC102.AB','GFGC120.AB','GGAL'])
	
	# Obtener el historico intradiario de GGAL en formato tick
	bsc.intraday_quotes(ticker='GGAL')

	# Obtener el historico intradiario de GGAL agrupado cada 5 minutos
	bsc.intraday_quotes(ticker='GGAL', timeframe=5)
	
	# Obtener el historico intradiario de GGAL agrupado cada 5 minutos 
	# ordenado en forma descendente
	bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0)

	# Obtener la ultima hora del historico intradiario de GGAL agrupado 
	# cada 5 minutos ordenado en forma descendente
	bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0, filter=60)

	# Obtener la ultima hora del historico intradiario de GGAL agrupado 
	# cada 5 minutos ordenado en forma descendente llenando el OHLC con 
	# el ultimo cierre si no existe cotizacion en el periodo
	bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0, filter=60, fillgap=1)


## Instalacion

Instalar ``bolsuite.python`` usando ``pip``:

    $ pip install git+https://github.com/crapher/bolsuite.python --upgrade --no-cache-dir


## Configuracion de Bolsuite

Para que ``bolsuite.python`` pueda acceder a la informacion provista por Bolsuite se debe agregar el tilde (circulo rojo) a la opcion **Permitir acceso a traves de WebAPI (JSON/XML)**

<img src="./images/bolsuite.configuration.png" width="500px">

**Nota:** En caso de que Bolsuite se ejecute en una maquina distinta a la que se esta utilizando ``bolsuite.python``, se debe sacar el tilde (circulo verde) de la opcion **Solamente permitir comunicaciones desde esta computadora (localhost)**


## Requerimientos

* [Python](https://www.python.org) >= 3.6+
* [Pandas](https://github.com/pydata/pandas) >= 1.0.0
* [Numpy](http://www.numpy.org) >= 1.18.1
* [requests](http://docs.python-requests.org/en/master/) >= 2.21.0


## Legal

**bolsuite.python** es distribuido bajo licencia **Apache Software License**. Para mas detalles mirar el archivo [LICENSE.txt](./LICENSE.txt).


## Contacto

Cualquier sugerencia que tengas, dejame un mensaje

**Diego A. Degese**

CHANGELOG.md
Change Log
===========

0.1.1
-------
- Se soluciono un problema que repetia la ultima cotizacion de nivel 2 cuando la siguiente cotizacion no tenia puntas.

0.1
-------
- Version Inicial (Beta)

bolsuite/connector.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Bolsuite Web API connector for Python
# https://github.com/crapher/bolsuite.python.git
#
# Copyright 2020 Diego Degese
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import requests as rq
import pandas as pd
import numpy as np
import json

class Connector():
    
    def __init__(self, base_url = 'http://localhost:6091'):
        self._base_url = base_url
    
    def _get_json(self, url):
        data = rq.get(url=url)
        
        if data.status_code == rq.codes.ok:
            return data.json()
                
        return json.loads("[]")

    def _get_dataframe(self, url):
        json = self._get_json(url)
        return pd.json_normalize(json)

    def _get_level2_dataframe(self, url):
        json = self._get_json(url)
        
        df = pd.DataFrame()

        def _get_level2_row(x, name, suffix, row):
            if type(x) is dict:
                for key in x:
                    _get_level2_row(x[key], name + key, suffix, row)
            elif type(x) is list:
                i = 1
                for item in x:
                    _get_level2_row(item, name, suffix + '.' + str(i), row)
                    i += 1
            else:
                if "Asks" in name:
                    name = name.replace('Asks', 'Ask')
                elif "Bids" in name:
                    name = name.replace('Bids', 'Bid')
                
                if "LastUpdate" in name and len(x) > 19:
                    x = x[:19] # Remove ms & Timezone data
                
                row[name + suffix] = x
                
            return row

        for item in json:
            row = _get_level2_row(item, name='', suffix='', row={})
            df = df.append(row, ignore_index=True)
            
        return df

    def _get_common_panel(self, url):
        df = self._get_dataframe(url)
        
        if not df.empty:
            df.drop(columns=['Source', 'Mode'], inplace=True)
            
            df.rename(columns={
                'Ask.Price': 'Ask',
                'Ask.Size': 'AskSize',
                'Bid.Size': 'BidSize',
                'Bid.Price': 'Bid',
                'LastQuote.DateTime': 'DateTime', 
                'LastQuote.Open': 'Open', 
                'LastQuote.High': 'High',
                'LastQuote.Low': 'Low', 
                'LastQuote.Close': 'Close', 
                'LastQuote.Volume': 'Volume'}, inplace=True)

            df['LastUpdate'] = df['LastUpdate'].str.slice(0, 19, 1)
            df['LastUpdate'] = pd.to_datetime(df['LastUpdate'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')
            df['DateTime'] = pd.to_datetime(df['DateTime'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')
            
            df.replace({'Settlement': {0: '', 1: 'Spot', 2: '24hs', 3: '48hs'}}, inplace=True)
            df.set_index(['Symbol', 'Settlement'], inplace=True)
            
            df.sort_index(inplace=True)
        
        return df
    
    def _get_options_panel(self, url):
        df = self._get_common_panel(url)
        
        if not df.empty:
            df['Expiration'] = pd.to_datetime(df['Expiration'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')

            if 'IntrisicValue' in df.columns: # Wrong column name to be fixed in a future Bolsuite version
                df.rename(columns={'IntrisicValue': 'IntrinsicValue'}, inplace=True)

            security_columns = [c for c in df.columns if c.lower()[:8] == 'security'] # Remove all the Security related columns
            df.drop(columns=security_columns, inplace=True)
        
        return df
        
    def _get_repos_panel(self, url):
        df = self._get_dataframe(url)
        
        if not df.empty:
            df.drop(columns=['Source', 'Mode'], inplace=True)
            
            df.rename(columns={
                'Ask.Value': 'AskValue',
                'Ask.Rate': 'AskRate',
                'Bid.Rate': 'BidRate',
                'Bid.Value': 'BidValue'}, inplace=True)

            df['LastUpdate'] = df['LastUpdate'].str.slice(0, 19, 1)
            df['LastUpdate'] = pd.to_datetime(df['LastUpdate'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')
            df['Expiration'] = pd.to_datetime(df['Expiration'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')
            
            df.set_index(['Days', 'Currency'], inplace=True)
            df.sort_index(inplace=True)
        
        return df
        
    def _get_quotes(self, url, isTick):
        df = self._get_dataframe(url)

        if not df.empty:
            df['DateTime'] = pd.to_datetime(df['DateTime'], format='%Y-%m-%dT%H:%M:%S',errors='coerce')
            
            if isTick:
                df.drop(columns=['Open', 'High', 'Low'], inplace=True)
                df.rename(columns={'Close': 'Price'}, inplace=True)
            
        return df

    def _get_level2_quotes(self, url):
        df = self._get_level2_dataframe(url)
        
        if not df.empty:
            df.drop(columns=['Source', 'Mode'], inplace=True)
            
            lastupdate_columns = [c for c in df.columns if 'lastupdate' in c.lower()]
            df[lastupdate_columns] = df[lastupdate_columns].apply(pd.to_datetime, format='%Y-%m-%dT%H:%M:%S', errors='coerce')

            df.replace({'Settlement': {0: '', 1: 'Spot', 2: '24hs', 3: '48hs'}}, inplace=True)
            df.set_index(['Symbol', 'Settlement'], inplace=True)

        return df
        
    def indices(self, ticker = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/indices/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/indices'.format(self._base_url)
        
        return self._get_common_panel(url)

    def bluechips(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
            settlement : str
                Opcional. Si se requiere obtener un settlement especifico (Valores Validos: Spot - 24hs - 48hs)  
                Default: None
        """
        if ticker is not None:
            url = '{}/api/bluechips/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/bluechips'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)
            
        return self._get_common_panel(url)

    def general_board(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
            settlement : str
                Opcional. Si se requiere obtener un settlement especifico (Valores Validos: Spot - 24hs - 48hs)  
                Default: None
        """
        if ticker is not None:
            url = '{}/api/generalboard/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/generalboard'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)

        return self._get_common_panel(url)

    def cedear_stocks(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/cedearstocks/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/cedearstocks'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)

        return self._get_common_panel(url)

    def government_bonds(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/governmentbonds/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/governmentbonds'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)

        return self._get_common_panel(url)

    def government_short_term_bonds(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/governmentshorttermbonds/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/governmentshorttermbonds'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)

        return self._get_common_panel(url)

    def options(self, underlying_asset):
        """
        :Parametros:
            underlying_asset : str
                El activo del que se necesitan las opciones
        """
        if underlying_asset is None or underlying_asset == '':
            raise Exception('Underlying asset is None or invalid')
            
        url = '{}/api/options/{}'.format(self._base_url, underlying_asset)
        
        return self._get_options_panel(url)

    def repos(self, days = None, currency = None):
        """
        :Parametros:
            days : int
                Opcional. Cantidad de dias a la expiracion
                Default: None
            currency : str
                Opcional. La moneda de la caucion (Monedas validas: ARS,USD)
                Default: None
        """
        if days is not None and currency is not None:
            url = '{}/api/repos/{}/{}'.format(self._base_url, days, currency)
        elif days is not None and currency is None:
            url = '{}/api/repos/{}'.format(self._base_url, days)
        elif days is None and currency is not None:
            url = '{}/api/repos/{}'.format(self._base_url, currency)
        else:
            url = '{}/api/repos'.format(self._base_url)
        
        return self._get_repos_panel(url)

    def corporate_bonds(self, ticker = None, settlement = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/corporatebonds/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/corporatebonds'.format(self._base_url)
        
        if settlement is not None:
            url = '{}?settlement={}'.format(url, settlement)

        return self._get_common_panel(url)

    def adrs(self, ticker = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/adrs/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/adrs'.format(self._base_url)
        
        return self._get_common_panel(url)

    def currency_pairs(self, ticker = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/currencypairs/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/currencypairs'.format(self._base_url)
        
        return self._get_common_panel(url)

    def commodities(self, ticker = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/commodities/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/commodities'.format(self._base_url)
        
        return self._get_common_panel(url)

    def personal_portfolio(self, ticker = None):
        """
        :Parametros:
            ticker : str
                Opcional. Si se requiere obtener un ticker especifico
                Default: None
        """
        if ticker is not None:
            url = '{}/api/personalportfolio/{}'.format(self._base_url, ticker)
        else:
            url = '{}/api/personalportfolio'.format(self._base_url)
        
        return self._get_common_panel(url)

    def level_2_quotes(self, tickers):
        """
        :Parametros:
            tickers : list
                La lista de tickers de la cual se necesita la profundidad de mercado
        """
        if tickers is None or len(tickers) == 0:
            raise Exception('Tickers is None or empty')
            
        url = '{}/api/level2?symbols={}'.format(self._base_url, '|'.join(tickers))

        return self._get_level2_quotes(url)

    def intraday_quotes(self, ticker, timeframe=None, sort=None, filter=None, fillgap=None):
        """
        :Parametros:
            ticker : str
                El ticker del cual se necesita la informacion historica
            timeframe : int
                Opcional. El intervalo en minutos que se va a usar para agrupar la informacion.
                0: tick - Mayor que 0: Cantidad de minutos
                Default: 0
            sort : int
                Opcional. La forma en que se va a ordenar la informacion retornada. (Valores Validos: 0: Mas nuevo a mas viejo - 1: Mas viejo a mas nuevo)                
                Default: 1
            filter : int
                Opcional. La cantidad de minutos hacia atras que se requiere la informacion.
                Default: 1800
            fillgap : int
                Opcional. En los timeframe mayores a 0 (1: Llenar los espacios que no hubo cotizaciones con la ultima cotizacion - 0: No agregar informacion)
                Default: 0
        """
        
        if ticker is None or ticker == '':
            raise Exception('Ticker is None or invalid')

        params = []
        if timeframe is not None:
            params.append('timeframe={}'.format(timeframe));
        if sort is not None:
            params.append('sort={}'.format(sort));
        if filter is not None:
            params.append('filter={}'.format(filter));
        if fillgap is not None:
            params.append('fillgap={}'.format(fillgap));
        
        if len(params) > 0:
            url = '{}/api/intraday/{}?{}'.format(self._base_url, ticker, '&'.join(params))
        else:
            url = '{}/api/intraday/{}'.format(self._base_url, ticker)

        return self._get_quotes(url, timeframe is None or timeframe == 0)


runtest.py
import bolsuite as bs

def test_connector():
    bsc = bs.Connector("http://10.211.55.3:6091")
    
    # Obtener todos los indices
    print(bsc.indices())
    
    # Obtener un indice filtrado por nombre
    print(bsc.indices(ticker='MERVAL'))
    
    # Obtener todas las acciones lideres
    print(bsc.bluechips())
    
    # Obtener todas las acciones lideres filtrada por vencimiento
    print(bsc.bluechips(settlement='Spot')) 
    
    # Obtener una accion del panel de lideres filtrada por nombre
    print(bsc.bluechips(ticker='PAMP'))
    
    # Obtener una accion del panel de lideres filtrada por nombre y vencimiento
    print(bsc.bluechips(ticker='PAMP', settlement='24hs'))
    
    # Obtener todas las acciones del panel general
    print(bsc.general_board())
    
    # Obtener todas las acciones del panel general filtrada por vencimiento
    print(bsc.general_board(settlement='48hs')) 
    
    # Obtener una accion del panel general filtrada por nombre
    print(bsc.general_board(ticker='LOMA'))
    
    # Obtener una accion del panel general filtrada por nombre y vencimiento
    print(bsc.general_board(ticker='LOMA', settlement='Spot'))
    
    # Obtener todos los cedears
    print(bsc.cedear_stocks())
    
    # Obtener todos los cedears filtrados por vencimiento
    print(bsc.cedear_stocks(settlement='24hs')) 
    
    # Obtener un cedear filtrado por nombre
    print(bsc.cedear_stocks(ticker='C'))
    
    # Obtener un cedear filtrado por nombre y vencimiento
    print(bsc.cedear_stocks(ticker='C', settlement='48hs'))
    
    # Obtener todos los bonos
    print(bsc.government_bonds())
    
    # Obtener todos los bonos filtrados por vencimiento
    print(bsc.government_bonds(settlement='Spot')) 
    
    # Obtener un bono filtrado por nombre
    print(bsc.government_bonds(ticker='AO20'))
    
    # Obtener un bono filtrado por nombre y vencimiento
    print(bsc.government_bonds(ticker='AO20', settlement='24hs'))
    
    # Obtener todas las letras
    print(bsc.government_short_term_bonds())
    
    # Obtener todas las letras filtradss por vencimiento
    print(bsc.government_short_term_bonds(settlement='48hs')) 
    
    # Obtener una letra filtrada por nombre
    print(bsc.government_short_term_bonds(ticker='S11M0'))
    
    # Obtener una letra filtrada por nombre y vencimiento
    print(bsc.government_short_term_bonds(ticker='S11M0', settlement='Spot'))
    
    # Obtener todas opciones de un subyacente
    print(bsc.options(underlying_asset='GGAL'))
    
    # Obtener todas las cauciones
    print(bsc.repos())
    
    # Obtener todas las cauciones filtradas por dia
    print(bsc.repos(days=7))
    
    # Obtener todas las cauciones filtradas por moneda
    print(bsc.repos(currency='ARS'))
    
    # Obtener una caucion filtrada por dia y por moneda
    print(bsc.repos(days=3, currency='USD'))
    
    # Obtener todas las obligaciones negociables
    print(bsc.corporate_bonds())
    
    # Obtener todas las obligaciones negociables por vencimiento
    print(bsc.corporate_bonds(settlement='24hs')) 
    
    # Obtener una obligacion negociable filtrada por nombre
    print(bsc.corporate_bonds(ticker='BFCPO'))
    
    # Obtener una obligacion negociable filtrado por nombre y vencimiento
    print(bsc.corporate_bonds(ticker='BFCPO', settlement='48hs'))
    
    # Obtener todos los ADRs
    print(bsc.adrs())
    
    # Obtener un ADR filtrado por nombre
    print(bsc.adrs(ticker='GGAL.O'))
    
    # Obtener todos los pares de divisas
    print(bsc.currency_pairs())
    
    # Obtener un par de divisas filtrado por nombre
    print(bsc.currency_pairs(ticker='USD-ARS'))
    
    # Obtener todos los pares de divisas
    print(bsc.commodities())
    
    # Obtener un ADR filtrado por nombre
    print(bsc.commodities(ticker='Brent Oil'))
    
    # Obtener todos los activos del panel personal
    print(bsc.personal_portfolio())
    
    # Obtener un activo del panel personal filtrado por nombre
    print(bsc.personal_portfolio(ticker='COME'))
    	
    # Obtener la profundidad de mercado de los activos seleccionados
    print(bsc.level_2_quotes(tickers=['GFGC102.AB','GFGC120.AB','GGAL']))
    
    # Obtener el historico intradiario de GGAL en formato tick
    print(bsc.intraday_quotes(ticker='GGAL'))
    
    # Obtener el historico intradiario de GGAL agrupado cada 5 minutos
    print(bsc.intraday_quotes(ticker='GGAL', timeframe=5))
    
    # Obtener el historico intradiario de GGAL agrupado cada 5 minutos ordenado en forma descendente
    print(bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0))
    
    # Obtener la ultima hora del historico intradiario de GGAL agrupado cada 5 minutos ordenado en forma descendente
    print(bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0, filter=60))
    
    # Obtener la ultima hora del historico intradiario de GGAL agrupado cada 5 minutos ordenado en forma descendente llenando el OHLC con el ultimo cierre si no existe cotizacion en el periodo
    print(bsc.intraday_quotes(ticker='GGAL', timeframe=5, sort=0, filter=60, fillgap=1))

if __name__ == "__main__":
    test_connector()